### PR TITLE
Category split

### PR DESCRIPTION
**Pull request by alschmiedt**
## The basics

- [ ] I branched from develop
- [ ] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves
1. Creates interfaces for a toolbox item, a selectable toolbox item and a collapsible toolbox item. 
2. Removes the Collapsible and Selectable toolbox item abstract classes
3. Splits the category into a `CollapsibleToolboxCategory` and a `ToolboxCategory`
4. Adds init methods to separate out creating the dom and updating the dom (expanding/hiding categories) Was also needed to fix a bug for hiding a nested category.
5. Moved getParent and getLevel into the ToolboxItem class. Since any toolbox item can technically be nested this needed to be on the toolbox item class. This also means that if someone extends the toolbox item class they don't have to deal with any of these methods.

### Test Coverage

Tested on:

### Additional Information

The refactor for the types is in the first commit. If you would like to split up when reviewing.

### CODE REVIEW

File: core/toolbox/collapsible_category.js [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+  }
+};
+
+/**
+ * Creates a toolbox item and adds it to the list of toolbox items.
+ * @param {!Blockly.utils.toolbox.ToolboxItemInfo} itemDef The information needed
+ *     to create a toolbox item.
+ * @private
+ */
+Blockly.CollapsibleToolboxCategory.prototype.createToolboxItem_ = function(itemDef) {
```
Comment (rachel-fenichel) [inline core/toolbox/collapsible_category.js:138]: @override ?

File: core/toolbox/toolbox.js [hunk 2]
```diff
@@ -392,12 +392,21 @@ Blockly.Toolbox.prototype.renderContents_ = function(toolboxDef) {
  * @private
  */
 Blockly.Toolbox.prototype.renderToolboxItem_ = function(toolboxItemDef, fragment) {
+  var kind = toolboxItemDef['kind'];
+
+  if (kind.toUpperCase() == 'CATEGORY' &&
```
Comment (rachel-fenichel) [inline core/toolbox/toolbox.js:397]: Add a comment explaining this extra conversion.

File: core/toolbox/toolbox.js [hunk 3]
```diff
@@ -531,7 +540,7 @@ Blockly.Toolbox.prototype.getWorkspace = function() {
 
 /**
  * Gets the selected item.
- * @return {?Blockly.ToolboxItem} The selected item, or null if no item is
+ * @return {?Blockly.IToolboxItem} The selected item, or null if no item is
```
Comment (rachel-fenichel) [inline core/toolbox/toolbox.js:543]: `ISelectableToolboxItem`. Same for the next couple methods.

File: core/toolbox/collapsible_category.js (hunk 1 above)
Comment (alschmiedt) [reply core/toolbox/collapsible_category.js:132]: I don't think this one needs an override, but I did miss it on a few others. So updated.

File: core/toolbox/toolbox.js (hunk 2 above)
Comment (alschmiedt) [reply core/toolbox/toolbox.js]: done

File: core/toolbox/toolbox.js (hunk 3 above)
Comment (alschmiedt) [reply core/toolbox/toolbox.js]: Done.

**Pull request merged**